base/system/winlogon/sas.c

```diff
@@ -285,6 +285,20 @@ PlaySoundRoutine(
     return Ret;
 }
 
+static
+BOOL
+IsFirstLogon(VOID)
+{
+    /* FIXME: All of this is a hack */
+    static BOOL bFirstLogon = TRUE;
+    if (bFirstLogon)
+    {
+        bFirstLogon = FALSE;
+        return TRUE;
+    }
+    return FALSE;
+}
```

Comment:
This function is a bit fragile in the sense that if it gets called twice in a row, the second call wouldn't return the expected result.

> Maybe checking first logon from WLSESSION or user token ?

In PlayLogonSoundThread, the `WLSESSION::UserToken` is already assigned to a non-NULL value corresponding to the user being logged in. Same for `WLSESSION::hProfileInfo`.
In the WLSESSION structure there's no existing variable that can help determining, at this point in time (in PlayLogonSoundThread), whether the current login is the first one or not:
- First, the `WLSESSION::LogonState` value automatically switches from `STATE_INIT` to `STATE_LOGGED_OFF` at startup [when the login dialog shows up](https://git.reactos.org/?p=reactos.git;a=blob;f=base/system/winlogon/winlogon.h;hb=92231340ec071bd7a461d8a941b4cd82b865bc81#l135), even before the first user tried to login.
- Everytime a user finished to log off, the `WLSESSION::UserToken` is [reset to NULL](https://git.reactos.org/?p=reactos.git;a=blob;f=base/system/winlogon/sas.c;h=c9675ba314de96cf4c4e99fd960848adea90dd42;hb=HEAD#l870). While it seems that `WLSESSION::hProfileInfo` is _NOT_ reset to NULL on logoff, I wouldn't advise you to base yourself on this fact as this can change in the future.
- `WLSESSION::LogonId` doesn't seem to be a good candidate either; I don't see it initialized to something anywhere in the code.

Reply:
> This function is a bit fragile in the sense that if it gets called twice in a row, the second call wouldn't return the expected result.

Yes, that's one of the reasons why it's a hack. It's definitely not thread-safe and should not be used for any other purposes.

Is our `WLSESSION` structure required to be fully compatible with Windows? If not then I have some ideas to implement this.
- Add a new `LastLogon` field that will be used to store the logon timestamp. Initialize it to `0` when the `WLSESSION` structure is allocated.
- Every time a user logon succeeded, store the timestamp (using the time from `GetTickCount` or `GetTickCount64`).
- Check for the first logon by checking `LastLogon != 0`.

Reply:
> This function is a bit fragile in the sense that if it gets called twice in a row, the second call wouldn't return the expected result.

Could using **Interlocked** fix that?

Reply:
No, here it's not about a synchronization problem betsween two (or more) separate threads.
What I meant is that, _**within one single function**_, if you do:
```c
do_something(IsFirstLogin() ? "this" : "that", ...);
do_something_else(IsFirstLogin() ? "whatnot" : "oops", ...);
```
the second invocation won't pass "whatnot" parameter but "oops".

Reply:
OK, so I have a question.
Is our `WLSESSION` structure required to be fully compatible with Windows? If not then I have some ideas to implement this.
- Add a new `LastLogon` field to the `WLSESSION` structure, initialize it to `0` at the time a `WLSESSION` is allocated.
- Every time a user logon succeeded, store the timestamp in `LastLogon` (using time from `GetTickCount` or `GetTickCount64` API call).
- Check for first logon by checking `LastLogon != 0`.

Reply:
>  Is our WLSESSION structure required to be fully compatible with Windows? If not then I have some ideas to implement this.

Actually WLSESSION is exclusively internal to our winlogon so it doesn't need to be (and is not already) compatible with Windows. So you can add whatever field suits you.
Your "LastLogon" also reminds me of a similar thing inside the msgina called [`LogonTime`](https://git.reactos.org/?p=reactos.git;a=blob;f=dll/win32/msgina/msgina.h;hb=92231340ec071bd7a461d8a941b4cd82b865bc81#l53) or the [`LastLogon`](https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&s=LastLog) stuff that is from the SAM dlls.
Not sure you need to do the GetTickCount thing; it may be useful to store a datestamp instead. For example like [there](https://git.reactos.org/?p=reactos.git;a=blob;f=dll/win32/samsrv/samrpc.c;hb=92231340ec071bd7a461d8a941b4cd82b865bc81#l7575)

Reply:
Maintaining a compatibility with Windows counterparts could be useful for experiments and cross-testing, but it's not a requirement. Can be done with some define.

Reply:
Thanks.
I will definitely replace the hack with this solution if it's possible. For this PR, I think the hack is good enough and could be kept as is, but I will rename the function and add more comments to be more precise.